[슬라이드 제작] 김민정 코드리뷰

minjeongss opened the pull request:
## 📂 진행도
- prev, next 버튼 누르면 무한 슬라이드 형식으로 작동
- pagination 버튼 누르면 해당 페이지의 슬라이드로 이동
- 좌측-우측 이동에 따른 애니메이션 구분

## 🤔 고민되는 부분
모든 이동을 queue의 형식으로 진행하다보니, `버튼을 누를 때마다 구조의 변화가 많아서 비효율적인 방법이 아닐까` 고민이 됩니다.

## ⚠️ 개선 필요한 부분
`handlePrevMove()`를 초기화할 때 실행하여 생기는 문제인데, 이 부분을 진행하지 않으면 슬라이드 2가 초기 화면에 보이게 됩니다. 이 부분을 일차적으로 해결하기 위해 슬라이드 2에서 1로 이동하도록 기존의 함수를 활용했습니다.

즉, 초기화를 진행할 때, 슬라이드 2에서 슬라이드 1로 이동하는 부분 수정이 필요합니다.

minjeongss committed:
feat: minjeongss slide 제작 업로드

zelkovaria commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+const $sliderDot = document.querySelector(".slider-dot");
+const $dots = $sliderDot.querySelectorAll("span");
+
+let isAnimating = false; // 애니메이션 진행중 여부
```
저는 진행중 여부를 고려하지 조차 않았었는데 왕꼼꼼하시네요..! 이런 사소함도 챙겨봐야겠어요👍🏻

zelkovaria commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  const $slider = document.querySelectorAll(".slider");
+
+  for (let i = 0; i < changeCount; i++) {
+    const lastElement = $sliderUl.lastElementChild.cloneNode(true);
```
제가 cloneNode랑 덜 친해서..ㅎㅎ 마지막 요소를 하나 복사해두고 맨 앞에 있던 슬라이드를 맨 뒤로 보내려는 의도라고 이해해도 되는걸까요...(?) 🥺

minjeongss replied:
네, 일부 맞습니다!
슬라이드 카드 하나 자체를 복사하는 방법이 무엇이 있을까 고민하다가 cloneNode를 사용하게 되었어요!

슬라이드가 1,2,3,4로 구성되어 있다고 가정해보겠습니다.
1. 4를 lastElement에 할당
2. 4를 ul에서 삭제
3. ul에서 1의 앞 위치에 4를 재배치

위의 순서로 이어진다고 보시면 될 것 같아요.
마지막 요소를 하나 복사하는 것은 맞지만, 맨 앞에 있던 슬라이드의 앞에 복사한 마지막 요소를 넣어주는 방식입니다.

zelkovaria commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  $sliderUl.style.transition = "none";
+  $sliderUl.style.transform = "translateX(-50%)";
+
+  requestAnimationFrame(() => {
```
애니메이션 최적화까지?? 외쳐 갓민정 👑
궁금한건 requestAnimationFrame으로 왜 두번이나 감싸주는건가요,,? 원리가 궁금해요 💭

minjeongss replied:
하핫 감사합니다 🤣
두 번 감싸서 작동하는 방식은 다음과 같습니다.

- 첫 번째 애니메이션 프레임에선 아래의 부분을 적용
```
$sliderUl.style.transition = "none";
$sliderUl.style.transform = "translateX(0)";
```
- 두 번째 애니메이션 프레임에선 아래의 부분을 적용
```
$sliderUl.style.transition = "transform 0.5s ease";
$sliderUl.style.transform = "translateX(-25%)";
```

두 차례의 적용 과정이 진행되기에 두 번 감싸주게 되는 것입니다.
이를 이론적인 측면에서 바라보면 브라우저가 리플로우와 리페인트 작업을 확실히 완료하고, 두 번째 호출 시점에서 트랜지션이 제대로 적용되도록 하기 위해서라고 합니다!

zelkovaria commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+
+const handlePageIndex = (item) => {
+  const $slider = document.querySelectorAll(".slider");
+  const $sliderIndex = Number($slider[1].innerText);
```
페이지네이션을 할 때 아예 InnerText 값을 바꿔볼 생각을 못했었는데 덕분에 새로운 방법 하나 터득하고 갑니당 💡

minjeongss replied:
해당 방법은 구성하는 레이아웃에 따라 바뀌겠지만, hidden으로 해도 되지 않을까 싶어 사용했습니다!
새로운 방법을 제시했다니, 뿌듯하네요!

zelkovaria commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  $dots[currentSliderIndex % $dots.length].classList.add("on");
+};
+
+$prevBtn.addEventListener("click", () => {
+  handlePrevMove(1);
+});
+$nextBtn.addEventListener("click", () => {
+  handleNextMove(1);
+});
+
+$sliderDot.addEventListener("click", (e) => {
+  let destinationDot = e.target;
+  handlePagination(destinationDot);
+});
```
함수 분리가 잘 되어있어서 가독성이 너무 좋은 것 같아요! ☺️

zelkovaria reviewed (commented):
애니메이션 중복 방지나 깔꼼한 함수 분리를 배워갈 수 있었습니다! 제가 부족한 메소드들도 발견할 수 있어서 너무 좋았습니다 🙇‍♀️ 수고하셨습니당 ㅎㅎ

ppyom commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+          isAnimating = false;
+          updatePagination();
+        },
+        { once: true }
+      );
+    });
+  });
+};
+
+const handleNextMove = (changeCount) => {
+  if (isAnimating) return;
+  isAnimating = true;
+
+  const $sliderUl = document.querySelector(".slider-ul");
+  const $slider = document.querySelectorAll(".slider");
+
+  for (let i = 0; i < changeCount; i++) {
+    const firstElement = $sliderUl.firstElementChild.cloneNode(true);
+    console.log(firstElement);
+    $sliderUl.removeChild($sliderUl.firstElementChild);
+    $sliderUl.appendChild(firstElement);
+  }
+  $sliderUl.style.transition = "none";
+  $sliderUl.style.transform = "translateX(0)";
+
+  requestAnimationFrame(() => {
+    requestAnimationFrame(() => {
+      $sliderUl.style.transition = "transform 0.5s ease";
+      $sliderUl.style.transform = "translateX(-25%)";
+      $sliderUl.addEventListener(
+        "transitionend",
+        () => {
+          isAnimating = false;
+          updatePagination();
+        },
+        { once: true }
+      );
+    });
+  });
+};
```
요부분..! 공통되는곳 함수로 분리할 수 있을 것 같습니당!!

그리고 다른 곳에서도 공통적으로 사용되는 `const $sliderUl = document.querySelector(".slider-ul");`은 전역으로 나와있어도 괜찮아보여요!!

딱 봤을 때 함수가 무슨 일을 하는지 슥 하고 읽혀서 정말 좋습니당!!👍

minjeongss replied:
스타일 부분을 분리하는걸 시도해볼게요!
$sliderUl은 전역으로 나오면 슬라이드 내용 변경이 반영이 올바르게 되지 않더군요...

혹시 그렇다면 $sliderUl 자체를 선언하는 것은 전역으로 빼고, 업데이트가 필요할 때마다 $sliderUl 자체에 재할당하는 방식으로 진행하는 것은 괜찮은 방법일까요? 🤔

(함수 구분에 대한 칭찬 정말 감사합니다. 춤추는 고래 상태입니다 🐳)

ppyom commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+
+const handlePageIndex = (item) => {
+  const $slider = document.querySelectorAll(".slider");
+  const $sliderIndex = Number($slider[1].innerText);
```
오!! 저도 이부분을 위치 옮길 생각뿐이였는데..! 새로운 방법 하나 알고갑니다👍

wilderif commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  if (isAnimating) return;
+  isAnimating = true;
+
+  const $sliderUl = document.querySelector(".slider-ul");
+  const $slider = document.querySelectorAll(".slider");
```
변수를 전역으로 두지 않고 필요한 부분에서 추출하는 것이 메모리 해제를 위해서 일까요??

minjeongss replied:
메모리 해제의 측면으로 사용한 것은 아니고, 슬라이드의 내용 자체를 변경하다 전역으로 두면 변경이 반영이 되지 않았습니다.. 그래서 사용할 때마다 추출했어요.

저 부분 수정이 필요하다고 생각을 했는데, 잊어버렸네요!
좋은 지적 감사합니다 (_ _) 👍

minjeongss replied:
$sliderUl과 $slider를 모두 지역에서 전역으로 위치를 변경하려고 시도를 했습니다.
다만, $sliderUl은 문제 없이 동작하나 $slider는 pagination 부분이 올바르게 동작하지 않아 지역으로 두었습니다!

wilderif replied:
DOM 변경이 일어나서 전역 변수를 그대로 유지하면 안되는 상황인가보네요. 다시 한 번 살펴보겠습니다!

minjeongss replied:
오 감사합니다! 😌

shlee9999 commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  const $slider = document.querySelectorAll(".slider");
+
+  for (let i = 0; i < changeCount; i++) {
+    const firstElement = $sliderUl.firstElementChild.cloneNode(true);
+    console.log(firstElement);
+    $sliderUl.removeChild($sliderUl.firstElementChild);
+    $sliderUl.appendChild(firstElement);
```
appendChild(firstElement)를 하시게 되면 firstElement가 마지막 자식 요소로 '이동'합니다! 추가되는 것이 아닙니다ㅎㅎ
clone하고 remove, append를 각각 따로 안해주셔도 돼요~
appendChild(firstElement)면 충분합니다

minjeongss replied:
와! 정말 몰랐어요.
클론하고 삭제하고 할당하는 3단계가 1단계로 축소될 수 있다니. 😮

좋은 지적 정말 감사합니다!

minjeongss replied:
해당 부분을 반영하다보니 handlePrevMove()에 대한 중복적인 부분도 축소했습니다! 👍

wilderif commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  $sliderUl.style.transition = "none";
+  $sliderUl.style.transform = "translateX(-50%)";
+
+  requestAnimationFrame(() => {
+    requestAnimationFrame(() => {
+      $sliderUl.style.transition = "transform 0.5s ease";
+      $sliderUl.style.transform = "translateX(-25%)";
+      $sliderUl.addEventListener(
+        "transitionend",
+        () => {
+          isAnimating = false;
+          updatePagination();
+        },
+        { once: true }
+      );
+    });
+  });
```
이 부분은 바로 이해가 되지 않아 따로 공부해보겠습니다. 좋은 방법 소개해 주셔서 감사합니다.

minjeongss replied:
저도 이번 슬라이드 구성을 하며 발견한 방법입니다!
웹 최적화와 애니메이션 타이밍을 보장할 수 있는 방법이어서 공부하시면 도움이 되리라 생각합니다 😊

저 부분을 사용하면서도 어려워서, 혹시 공부하고 정리하신 내용 공유해주시면 큰~도움이 될 것 같습니다.

ppyom reviewed (commented):
전체적으로 뭔가 코드 읽을 사람을 위해 차근차근 작성되어있는 느낌이 있어 읽는데 불편함 없이 좋았습니다!! 👍

wilderif commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  handleNextMove(1);
+});
+
+$sliderDot.addEventListener("click", (e) => {
+  let destinationDot = e.target;
+  handlePagination(destinationDot);
+});
```
이벤트를 모든 점에 붙이지 않고 처리한 점이 오늘 아침 RBF 시간에 설명해주신 부분 잘 활용하신 것 같아요!

minjeongss replied:
RBF 시간에 이벤트 위임과 이벤트 버블링을 설명하고, 수업이 시작되기 전에 이벤트 위임을 적용한 방식으로 변경했어요.
이 부분을 알아봐주시다니~ 감동입니다. 🤩

jinhyukSeo777 commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  });
+};
+
+const handlePageIndex = (item) => {
```
handlePageIndex라는 이름이 처음 읽을땐 잘 와닫지 않는것 같아요..! pageIndex를 바꾸는 함수인가? 싶어가지고 getIndexDiff나 다른 명확한 이름으로 하는건 어떨까요?

minjeongss replied:
오! 생각하지 못했는데, get으로 바꾸는게 더 기능이 명확하게 보일 것 같아요.
리팩토링하면서 해당 부분 함수 이름 getIndexDiff로 변경했습니다.

세심한 관찰 감사드려요! 😊

jinhyukSeo777 commented on minjeongss/slide.js:
```diff
@@ -0,0 +1,114 @@
+  const $sliderIndex = Number($slider[1].innerText);
+  const $paginationIndex = Number(item.classList[0][3]);
+  console.log("slider", $sliderIndex, "   /pagination", $paginationIndex);
+  let diffIndex = $sliderIndex - $paginationIndex;
```
diffIndex는 diff를 가르키는 index처럼 읽혀서 차라리 indexDiff나 다른거로 바꾸는건 어떨까요?

minjeongss replied:
sliderIndex나 paginationIndex처럼 앞에 수식하고 마지막에 정체가 Index이다!라고 나타내고 싶어서 해당 방식을 사용했어요.
저는 개인적으로 변수 순서명이 통일되지 못하면, 자신이 짠 코드여서 정체를 파악하기가 어렵더군요.. 😅

진혁님은 indexDiff로 변수명 짓는걸 선호하시나요? 
혹시 그러면 sliderIndex와 paginationIndex 부분은 어떤 방식으로 명명하실 것 같나요??

jinhyukSeo777 replied:
sliderIndex와 paginationIndex는 저도 똑같이 사용할 것 같아요! 보통 index라고 하면 0, 1, 2처럼 0부터 시작해서 요소의 순서를 나타내는데 diffIndex 는 +도 -도 될 수 있는 index간의 차이인거 같아서 순서가 아니라 차이 값이면 diff가 뒤로오면 좋지 않을까 했습니다...!
사실 본인한테 잘 와닫으면 그게 최고긴하죠😊

minjeongss replied:
오! 어떤 측면에서 커멘트를 남겨주신지 이해했어요.
인덱스의 특성을 고려하라는 조언이었군요...
해당 변수는 리팩토링할 때 diffIndex로 사용하긴 했지만, 그런 부분도 다음엔 곡 고려해볼게요! 감사해요~!

minjeongss committed:
feat: 피드백 후 코드 수정

- $sliderUl 전역으로 재배치
- handlePrevMove, handleNextMove에서 복제-삭제-할당이 아닌 할당만 진행
- 반복되는 CSS 함수로 구현
- 초기화 함수 명시화
